DIPs/1NNN-JH-TG-NT.md

```diff
@@ -0,0 +1,483 @@
+redefine the comma operator syntax.
+
+```d
+auto t = (1, 2); // tuple literal
+int x, y;
+(x, y) = t;
+assert(x == 1);
+assert(y == 2);
+```
+Note: Using the result of a comma expression is already an error.
+
+
+### Unpacking function literal parameters
+A function parameter must have a type, so function parameters cannot be unpacked without
+tuple types, which are beyond the scope of this DIP. However, a function literal template
+with an untyped unpack declaration parameter is supported:
+```d
+alias dg = ((x,y), z) => writeln(x," ",y," ",z);
+dg(tuple(1, 2), 3); // "1 2 3\n"
+```
+
+`((x,y), z){ ... }` is lowered to `(__arg0, z){ auto (x,y) = __arg0; ...}`. In general, it preserves and copies storage classes from
+the parameter to the unpack declaration in the function body.
+
+`ref` storage class is copied outwards, e.g. if you do `((ref a,b), c){ ... }` that gets lowered to
+`(ref __arg0, c){ (ref a,auto b) = __arg0; ... }`. 
+
+`out` can be applied to the whole unpack parameter, but not to individual fields. `auto ref` and `lazy` cannot
+be applied on an unpacking parameter list.
+
+Note: `out` unpacking parameters are not supported by the current implementation, but may be added at a future date.
+
+
+### Grammar changes
+The following grammar additions will be sufficient.
+(Note that many grammar rules are repeated due to (existing) distinct restrictions on what `StorageClasses` are allowed; this can make the proposed grammar changes look more profound than they are.)
+
+#### Unpacking declarations
+See <https://dlang.org/spec/declaration.html#VarDeclarations>.
+```diff
+  VarDeclarations:
+      ...
+      AutoDeclaration
++     StorageClasses[opt] TupleDeclarators ;
+
++ TupleDeclarators:
++     TupleDeclarator
++     TupleDeclarator , TupleDeclarators
+
++ TupleDeclarator:
++     ( TupleDeclarators2 ) = Initializer
+
++ TupleDeclarators2:
++     TupleDeclarator2 ,
++     TupleDeclarator2 , TupleDeclarator2
++     TupleDeclarator2 , TupleDeclarators2
+
++ TupleDeclarator2:
++     StorageClasses[opt] Identifier
++     StorageClasses[opt] BasicType TypeSuffixes[opt] Identifier
++     StorageClasses[opt] ( TupleDeclarators2 )
+```
+
+#### Unpacking `foreach` elements
+(Note that `ForeachTupleDeclarators2` is the same as `TupleDeclarators2` except that the latter allows all `StorageClasses`, while the former allows only `ForeachTypeAttributes`.)
+
+See <https://dlang.org/spec/statement.html#ForeachType>.
+```diff
+  ForeachType:
+      ...
+      ForeachTypeAttributes[opt] alias Identifier
++     ForeachTypeAttributes[opt] ( ForeachTupleDeclarators2 )
+
++ ForeachTupleDeclarators2:
++     ForeachTupleDeclarator2 ,
++     ForeachTupleDeclarator2 , ForeachTupleDeclarator2
++     ForeachTupleDeclarator2 , ForeachTupleDeclarators2
+
++ ForeachTupleDeclarator2:
++     ForeachTypeAttributes[opt] Identifier
++     ForeachTypeAttributes[opt] BasicType TypeSuffixes[opt] Identifier
++     ForeachTypeAttributes[opt] ( ForeachTupleDeclarators2 )
```

Comment:
Here, `ForeachTupleDeclarator(s)2` is a quite uninformative name. I suggest `ForeachTupleInnerDeclarator(s)`. (This doesn’t change what the grammar defines.)
```suggestion
+     ForeachTypeAttributes[opt] ( ForeachTupleDeclarators2 )

+ ForeachTupleInnerDeclarators:
+     ForeachTupleInnerDeclarator ,
+     ForeachTupleInnerDeclarator , ForeachTupleInnerDeclarator
+     ForeachTupleInnerDeclarator , ForeachTupleInnerDeclarators

+ ForeachTupleInnerDeclarator:
+     ForeachTypeAttributes[opt] Identifier
+     ForeachTypeAttributes[opt] BasicType TypeSuffixes[opt] Identifier
+     ForeachTypeAttributes[opt] ( ForeachTupleInnerDeclarators )
```